Fix part of #21970: Domain object and validate method for ExplorationVoiceArtistLinkModel.

Ashu463 opened the pull request:
## Overview

1. This PR fixes part of #21970 
2. This PR does the following: Creates a domain object for ExplorationVoiceArtistLinkModel, and hence validate method for the same. This PR also also makes sure that this domain object is used everywhere in the codebase in place of its models across domain layer. 

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [X] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [X] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [X] I have written tests for my code.
- [X] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [X] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

Changes that I made on this PR, is verified by all passed backend_tests and updated beam jobs shown in below ss: 

- Backend test could be seen in CI workflow, no need to attach any ss for any specific test class.
- For `maunal_voice_artist_name_job`: 

![image](https://github.com/user-attachments/assets/ae33e74c-05f1-4194-9477-eb7adea48c65)

- For `voiceover_migration_job`:

![image](https://github.com/user-attachments/assets/c7a459be-268f-40bb-885c-429470acc969)


## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Ashu463 committed:
feat: added domain object for voiceOverArtistsLinkModel, created validate method, updated voiceover_services file to use newly created domain object

Ashu463 committed:
fix: mypy, lint errors and some typos

Ashu463 committed:
fix: other validation jobs are now using newly created domain obj

Ashu463 committed:
fix: corresponding test cases are fixed

Ashu463 commented:
@ankita240796 PTAL. 
Thanks.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
+        self.exp_id = exp_id
+        self.content_id_to_voiceovers_mapping = content_id_to_voiceovers_mapping
+
+    @classmethod
+    def from_dict(
+        cls, exploration_voiceArtists: ExplorationVoiceArtistsLinkDict
+    ) -> ExplorationVoiceArtistsLink:
+        """Returns a explorationVoiceArtistsLink domain object from a dict.
+
+        Args:
+            exploration_voiceArtists: Dict. Dict represents object
+                of Exploration voice artists link object.
+
+        Returns:
+            ExplorationVoiceArtistsLink: Object.
+            Exploration voice artists link object instance.
+        """
+        return cls(
+            exploration_voiceArtists['exp_id'],
+            exploration_voiceArtists['content_id_to_voiceovers_mapping']
+        )
+
+    def to_dict(self) -> ExplorationVoiceArtistsLinkDict:
+        """Returns a dict representing a exploration voice artist link
+        data domain object.
+
+        Returns:
+            ExplorationVoiceArtistsLinkDict. Dict. A dict, 
+            mapping content ID to respective voice over artist.
+        """
+
+        return {
+            'exp_id': self.exp_id,
+            'content_id_to_voiceovers_mapping': (
+                self.content_id_to_voiceovers_mapping)
+        }
```
Is there any usecase for these methods in the current design? If not, let's remove.

Ashu463 replied:
Currently these aren't used yet though removed, I took reference from another domain objects to built similar architecture. But what if someone needs this in future ?

ankita240796 replied:
We add it when it's needed. Having code which is not required assuming it will be required in future is not a good idea. If we can think of usecases on why this is needed or will be needed, we can list that down first and add it back.

ankita240796 replied:
Also, do all domain objects have `to_dict` and `from_dict` functionality?

Ashu463 replied:
@ankita240796 Yup, as much as I can see almost all domain objects have to_dict and from_dict methods.

ankita240796 replied:
How did you check that?

I see some examples which don't:

- https://github.com/oppia/oppia/blob/develop/core/domain/activity_domain.py#L101
- https://github.com/oppia/oppia/blob/develop/core/domain/auth_domain.py#L116 (doesn't have from_dict)
- All objects in https://github.com/oppia/oppia/blob/develop/core/domain/blog_statistics_domain.py

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
             automated_voiceovers_audio_offsets_msecs={})
+
+
+class ExplorationVoiceArtistsLinkDict(TypedDict):
+    """Dict type for Exploration Voice Artists object."""
+
+    exp_id: str
+    content_id_to_voiceovers_mapping: Dict[
+        str, Dict[str, Tuple[str, state_domain.VoiceoverDict]]]
```
Ditto.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
+
+    def __init__(
+            self,
+            exp_id: str,
```
Is this required anywhere?

Ashu463 replied:
Yeah, this constructor is called whenever someone intializes ExplorationVoiceArtistsLink domain object.

ankita240796 replied:
I meant `exp_id`.

ankita240796 replied:
I meant `exp_id`.

Ashu463 replied:
Yeah, exp_id is also required to create/update contents within `content_id_to_voiceovers_mapping`. 

See `create_exploration_voice_artists_link_model_instance` and `update_exploration_voice_artist_link_model` for reference.

ankita240796 replied:
Hmm, so isn't it only required to create the model but not the domain object? Is there any caller of domain object which will need access to `exp_id`?

What I mean is if the only purpose of `exp_id` is in model to get model and save model, we should not have it in domain object. Does that make sense?

Ashu463 replied:
I think you are right domain object shouldn't hold any attrs if it only needed in performing actions over models. Left the similar cmnt [here](https://github.com/oppia/oppia/pull/22186#discussion_r2009017265) in support of removing exp_id.

ankita240796 replied:
Sgtm, makes sense.

ankita240796 replied:
Is this updated?

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
+
+        Args: 
+            exp_id: str. ID of exploration.
+            content_id_to_voiceovers_mapping: Dict. A dict consisting 
+                of content IDs as key and other dicts as value.
+                This other dict consists of a bunch of languages
+                (in which voiceovers of content is made) as keys
+                and a two-tuple as values. First value of tuple
+                is voice artist ID and second value is VoiceOverDict.
```
Let's copy over the comment from [here](https://github.com/oppia/oppia/blob/7e06003db0f867ea94ee07af459f6a3b616bcb00/core/storage/voiceover/gae_models.py#L393C5-L396C29).

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
+            )
+        # Validating content within content_id_to_voiceovers_mapping dict.
+        for content_id, nested_dict in (
+            self.content_id_to_voiceovers_mapping.items()):
```
Fix indent.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
+                # Validating langauge.
+                if not isinstance(lang, str):
+                    raise utils.ValidationError(
+                    'Expected language to be string'
+                    'But recieved: %s' % lang
```
Fix indent.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
+                    % nested_dict
+                )
+            # Validating dicts inside each content.
+            for lang, (artist_id, voiceover_dict) in nested_dict.items():
```
language_code

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
+            # Validating dicts inside each content.
+            for lang, (artist_id, voiceover_dict) in nested_dict.items():
+                # Validating langauge.
+                if not isinstance(lang, str):
```
What about validating the language code?

Ashu463 replied:
Ah I just forgot that, added validation for it now, PTAL. @ankita240796 one thing I'm assuming that language code is coming without its accent. Lemme explain, there are two things language codes (en, hi, etc.) and other is langauge accent code (en-US, en-CA, etc.), on observing closely to test cases(that were built prior to my changes) of voiceover_domain I got this assumption.

ankita240796 replied:
Hmm, I am not sure about that, based on the comment on the model fields, what you assume seems to be the case. But let us double check. Looping in @oppia/core-reviewers who own the code for confirming.

seanlip replied:
Please check this with @Nik-09, thanks.

Nik-09 replied:
Yes, @Ashu463 you are correct, there are two concepts: **language_code** and **language_accent_code**. Here, language code is concerned. Also, I would encourage you to use a more meaningful name [here](https://github.com/oppia/oppia/pull/22186/files#diff-9d52012a4ea7046fbd3e16f8be487ac0bbf2cfa1bae01b0850336ac918e67eabR407) instead of nested_dict.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
+                )
+                if not isinstance(artist_id, str):
+                    raise utils.ValidationError(
+                    'Expected artist id to be string'
+                    'But recieved: %s' % artist_id
```
Fix indent.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
+                    'Expected artist id to be string'
+                    'But recieved: %s' % artist_id
+                )
+                # Validating voiceOverDict.
```
Remove, not required.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,160 @@ def create_empty(
+                        % voiceover_dict
+                    )
+
+                expected_keys = {
```
Does `state_domain.Voiceover` have validating for it? If yes, can we obtain `Voiceover` domain object from this dict and invoke the existing validate method?

Ashu463 replied:
@ankita240796 yep, state_domain.Voiceover is validating in a pretty good way, and now I'm using that for validation with the help of this `state_domain.Voiceover.from_dict(voiceover_dict).validate()`

ankita240796 replied:
Thanks, but why do we need the expected keys validation here. Does it not happen in `state_domain`?

ankita240796 replied:
@Ashu463 can you please respond to each comment and not just react since it will clarify things where a clarification is required? Please follow https://github.com/oppia/oppia/wiki/Rules-for-making-PRs#step-5-address-review-comments-until-all-reviewers-approve.

I am not sure what the react denotes above but I am guessing you mean expected keys validation is not happening in `state_domain`. If that's the case, let's add it there and not here.

Ashu463 replied:
@ankita240796 I responded to this on my PTAL [comment](https://github.com/oppia/oppia/pull/22186#issuecomment-2744592034), sry for that will reply to cmnt threads only now. And yeah there were no any implementation regarding expected_keys in state_domain so I wrote from my side.

ankita240796 replied:
Thanks @Ashu463, I do think that any validation we implement should not be scattered at different places as that makes it harder to read and access for any other developers. So, let's add missing validation for `Voiceover` in `state_domain` itself instead of having some bits and pieces here.

Ashu463 replied:
@ankita240796 one thing just clicked in my mind, if there were extra fields sent over state_domain.Voiceover model from_dict would only pick needed ones, and if some fields were missing then it will throw a keyError for missing fields so there is no need of validation for expected_keys if sent to state_domain.Voiceover directly.

ankita240796 replied:
Yes, that makes sense. You can handle the key errors raised in a try catch block and generate validation errors accordingly.

Ashu463 replied:
Hmm, Done.

lkbhitesh07 replied:
As per the above discussion, I am guessing that the keys validation should be removed?

Ashu463 replied:
That's my bad, actually I thought it had been removed. Removed it now.

ankita240796 commented on core/jobs/batch_jobs/manual_voice_artist_name_job.py:
Let's revert this, any jobs should use prod models.

Ashu463 replied:
@ankita240796 reverted this, that was my bad I missed that this job is making call to model directly.

ankita240796 commented on core/jobs/batch_jobs/voiceover_migration_job.py:
Ditto.

Ashu463 replied:
Same for here.

ankita240796 reviewed (changes requested):
Thanks @Ashu463, took a first pass, PTAL.

Ashu463 committed:
fix: iteration#1 changes

Ashu463 commented:
@ankita240796 PTAL now. I'd made changes according to your suggestions and replied to your threads where a discussion is needed.

Thanks.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+            )
+        # Validating content within content_id_to_voiceovers_mapping dict.
+        for content_id, nested_dict in (
+        self.content_id_to_voiceovers_mapping.items()):
```
Fix indent.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+                )
+            # Validating dicts inside each content.
+            for language_code, (artist_id, voiceover_dict) in (
+                nested_dict.items()):
```
Fix indent.

ankita240796 replied:
Fix indent.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+                'Expected exploration id to be string, recieved: %s.'
+                % self.exp_id
+            )
+        if self.exp_id == '':
```
`not self.exp_id`?

Ashu463 replied:
Hey I'm raising error if exp_id is empty it will pass all other cases with non empty strings, I think you're supposing just opposite case. Right ?

ankita240796 replied:
I meant instead of `if self.exp_id == ''`, can we do `if not self.exp_id`?

Ashu463 replied:
Ahh fine, yeah that could be also done. Fixed for each similar cases.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+            )
+        if self.exp_id == '':
+            raise utils.ValidationError(
+                'Exploration ID should not be empty.'
```
Is the break needed? Will it not fit on the previous line in the allowed character limit?

Ashu463 replied:
Fixed it.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+                nested_dict.items()):
+                if not isinstance(language_code, str):
+                    raise utils.ValidationError(
+                    'Expected language to be string recieved: %s'
```
Fix indent.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+                    'Expected language to be string recieved: %s'
+                    % language_code
+                )
+                allowed_language_codes = [language['id'] for language in (
+                constants.SUPPORTED_AUDIO_LANGUAGES)]
+                if language_code not in allowed_language_codes:
+                    raise utils.ValidationError(
+                        'Expected language code to be one from supported '
+                        'audio languages recieved: %s' % language_code
+                    )
```
There is a function in utils to do this, let's use that to avoid repetition of same logic.

ankita240796 replied:
Good catch on using supported audio languages and not just all langauges!

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+                    )
+                if not isinstance(artist_id, str):
+                    raise utils.ValidationError(
+                    'Expected artist id to be string recieved: %s'
```
Fix indent.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+                )
+                if not isinstance(voiceover_dict, dict):
+                    raise utils.ValidationError(
+                    'Expected voiceover_dict to be dict recieved: %s'
```
Fix indent.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+            )
+        # Validating content within content_id_to_voiceovers_mapping dict.
+        for content_id, nested_dict in (
+        self.content_id_to_voiceovers_mapping.items()):
```
Fix indent.

ankita240796 replied:
This is still not fixed, you need to indent when you break after a bracket.
```suggestion
        for content_id, nested_dict in (
            self.content_id_to_voiceovers_mapping.items()):
```

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+                'Expected content_id_to_voiceovers_mapping to be Dict'
+                'But recieved: %s' % self.content_id_to_voiceovers_mapping
+            )
+        # Validating content within content_id_to_voiceovers_mapping dict.
```
Remove this comment, it's already clear.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+            if not isinstance(content_id, str):
+                raise utils.ValidationError(
+                    'Expected content id to be string, recieved: %s.'
+                    % content_id
```
Add bracket and indent.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+            if not isinstance(nested_dict, dict):
+                raise utils.ValidationError(
+                    'Expected nested dict to be Dictionary, recieved: %s.'
+                    % nested_dict
```
Ditto.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+                        'Expected language code to be one from supported '
+                        'audio languages recieved: %s' % language_code
+                    )
+                if not isinstance(artist_id, str):
```
What about `artist_id` being empty string?

Ashu463 replied:
Added validation for it.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +317,100 @@ def create_empty(
+        # Validating content within content_id_to_voiceovers_mapping dict.
+        for content_id, nested_dict in (
+        self.content_id_to_voiceovers_mapping.items()):
+            if not isinstance(content_id, str):
```
What about `content_id` being empty string?

Ashu463 replied:
Added validation.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+    def setUp(self) -> None:
+        super().setUp()
+
+        self.dummy_voiceover_dict: state_domain.VoiceoverDict = {
```
Remove dummy. Just `voiceover_dict` is fine.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+
+
+class ExplorationVoiceArtistsLinkUnitTests(test_utils.GenericTestBase):
+    """Test for ExplorationVoiceArtistsLink domain class."""
```
domain class -> domain object

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+                content_id_to_voiceovers_mapping=(
+                    self.content_id_to_voiceovers_mapping))
+            )
+        self.exp_voice_artists_link.validate()
```
Why is this required in setup?

Ashu463 replied:
Just learnt from previous one, updated it PTAL at new one.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+        self.exp_voice_artists_link.validate()
+
+    def test_validate_exp_id(self) -> None:
+        # Test exp_id must be a string.
```
Remove.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+            self.exp_voice_artists_link.exp_id = 123  # type: ignore[assignment]
+            self.exp_voice_artists_link.validate()
+
+        # Reset for next test.
```
Remove - avoid adding comments for sefl explanatory stuff.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+        # Reset for next test.
+        self.exp_voice_artists_link.exp_id = 'exp_id_1'
+
+        # Test exp_id should not be empty.
```
Remove.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+            self.exp_voice_artists_link.validate()
+
+        # Reset for next test.
+        self.exp_voice_artists_link.exp_id = 'exp_id_1'
```
Break this down, don't test multiple test functionalities in a single test function. Keep it as granular as possible.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+        self.exp_voice_artists_link.exp_id = 'exp_id_1'
+
+        # Test exp_id should not be empty.
+        with self.assertRaisesRegex(
```
Have a separate test for this.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+            utils.ValidationError,
+            'Expected content_id_to_voiceovers_mapping to be Dict'
+        ):
+            self.exp_voice_artists_link.content_id_to_voiceovers_mapping = 'not_a_dict'  # type: ignore[assignment]
```
Please create new domain objects in each test rather than having one in setup. For tests, prefer damp over dry, refer: https://enterprisecraftsmanship.com/posts/dry-damp-unit-tests/

ankita240796 replied:
Please make sure you fix it throughout all the tests.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+            self.exp_voice_artists_link.validate()
+
+    def test_validate_content_id(self) -> None:
+        # Test content_id must be a string.
```
Remove - make function name descriptive if required instead.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+            self.exp_voice_artists_link.validate()
+
+    def test_validate_nested_dict(self) -> None:
+        # Test nested dict must be a dictionary.
```
Ditto.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+            self.exp_voice_artists_link.validate()
+
+    def test_validate_language_code(self) -> None:
+        # Test language must be a string.
```
Remove.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+            self.exp_voice_artists_link.validate()
+
+    def test_validate_language_code_present_in_supported_lang(self) -> None:
+        # Test language code must be from one of predefined supported audio
```
Remove.

ankita240796 commented on core/domain/voiceover_domain_test.py:
```diff
@@ -394,3 +394,381 @@ def test_is_both_voiceovers_empty_should_return_successfully(self) -> None:
+            self.exp_voice_artists_link.validate()
+
+    def test_validate_artist_id(self) -> None:
+        # Test artist_id must be a string.
```
Remove. Please do this here and everywhere else required.

ankita240796 commented on core/domain/voiceover_services.py:
```diff
@@ -493,6 +493,58 @@ def get_autogeneratable_language_accent_list() -> Dict[str, Dict[str, str]]:
+        model.content_id_to_voiceovers_mapping = (
+            exploration_voice_artists_link.content_id_to_voiceovers_mapping)
+
+    model.update_timestamps()
```
What does `update_timestamps` do?

Ashu463 replied:
I added this coz I see it here, so felt necessary to have: 
https://github.com/oppia/oppia/blob/c427e3aaa75f97bfc4e915ed2db9022ea752fbc3/core/domain/voiceover_services.py#L824

Although this isn't doing anything behind the scenes, as far as I could rememeber, implementation of update_timestamps() and put() is listed in TODO of several other models.

ankita240796 replied:
It does not answer the question though. The reason should not be that some other code has it since it is unclear as to why the code you added is dependent on the other code and further does not demonstrate your understanding of the code. So, here is how I would advice you to answer above queries:

- Go to the def for `update_timestamps`: https://github.com/oppia/oppia/blob/develop/core/storage/base_model/gae_models.py#L404 - it sets the `created_on` and `last_udpated` fields of the model.
- This is done to ensure the time fields are populated correctly, hence invoked before `put` which saves the model in datastore.

My question was coming from the place of assuming that `put` should ideally `update_timestamps` and we should not need an extra call to `update_timestamps`. I am not sure on why that is not the case but I can confirm that it is the case after checking implementation of `put` where no call is made to update timestamps. @seanlip I assumed ndb put should handle timestamps correcty, checking with you if there is some knowledge I am missing here. 

> implementation of update_timestamps() and put() is listed in TODO of several other models.

Similarly, when you say the above explain what these TODOs are, provide links and elaborate on why they are relevant here.

Ashu463 replied:
Sorry @ankita240796  for such vague reply, now for put method here is something I got, instead of using put() method we should prefer to use put_for_human() which updates timestamps accordingly, which even solved your question of calling update_timestamps() twice.

https://github.com/oppia/oppia/blob/770936014c68d791c1ccc33fe9df0fec268779ca/core/storage/base_model/gae_models.py#L603

And excuse me for saying that put() method is in TODO, actually according to codebase it is implemented in derived class, see here : https://github.com/oppia/oppia/blob/770936014c68d791c1ccc33fe9df0fec268779ca/core/storage/base_model/gae_models.py#L1256

ankita240796 replied:
> instead of using put() method we should prefer to use put_for_human() which updates timestamps accordingly

Hmm, I do not think we should do that - that contradicts the function definition as `put_for_human` is used when someone manually saves a model, right?

It is okay to use `update_timestamps` as long as we know why we are using it. Does that make sense?

Ashu463 replied:
Fine, just a quick note: I am only updating these changes with the model I'm working on `ExplorationVoiceArtistsLinkModel`.

seanlip replied:
> My question was coming from the place of assuming that put should ideally update_timestamps and we should not need an extra call to update_timestamps. I am not sure on why that is not the case but I can confirm that it is the case after checking implementation of put where no call is made to update timestamps. @seanlip I assumed ndb put should handle timestamps correcty, checking with you if there is some knowledge I am missing here.

Generally speaking we only want to update the last-updated timestamp if the action really changes the underlying data (as opposed to e.g. a schema migration done by a bot, since that doesn't change the semantics of what's stored in the model and it wlil result in the last-updated timestamps being the same for all models when e.g. a cron job runs). So that's why put() shouldn't necessarily update timestamps directly.

Does what we're doing here fall into the "substantive edit" case or not? If the former then updating the last-updated timestamp is appropriate; if the latter then it isn't.

ankita240796 replied:
Thanks for clarifying @seanlip, that makes sense. What we are doing here is a "substantive edit" case as we are updating a model field.

ankita240796 commented on core/domain/voiceover_services.py:
```diff
@@ -514,6 +566,14 @@ def get_all_voice_artist_language_accent_mapping() -> Dict[str, Dict[str, str]]:
             voiceover_models.ExplorationVoiceArtistsLinkModel.get_all().fetch()
         )
 
+    exploration_voice_artist_links = [
+        voiceover_domain.ExplorationVoiceArtistsLink(
+            exp_id=model.id,
+            content_id_to_voiceovers_mapping=(
+                model.content_id_to_voiceovers_mapping)
+        ) for model in exploration_voice_artist_link_models
+    ]
+
```
This and below is not useful. There is no purpose of domain object if we just use it internally in `voiceover_services` to get the field which we could have gotten from the model instead. Here is what I would suggest:

- Have functions which can be invoked on the domain object to obtain `voice_artist_language_accent_mapping`
- Return domain objects from `voiceover_services` instead of returning mappings and let the callers further invoke required methods on domain objects.

Ashu463 replied:
@ankita240796 I'd replaced this whole method with totally new approach(your suggestions). This method is replaced by following set of logic everywhere it is used: 
- Make a call to get `get_voice_artist_metadata_mapping`. 
- Then a call to get `get_exploration_voice_artist_links`. 
- And then finally calling the main logic to get mapping from newly created method( get_voice_artist_language_mappings) inside domain object. 

And I did this at three place `voiceover_services_test`, `voiceover`, and `voiceover_test` where this whole logic is used.

ankita240796 reviewed (changes requested):
Thanks @Ashu463, took another pass, PTAL.

Ashu463 committed:
Merge branch 'develop' into issue#21970-ExplorationVoiceArtistsLink

Ashu463 committed:
fix: iteration#2 suggestions

Ashu463 commented:
@ankita240796 PTAL, Fixes in this iteration: 
- I'd updated all test cases to create domain objects within them and hence validating. 
- And yes I added validation for expected_keys coz state_domain isn't validating on this criterion.
- Updated way of using `get_all_voice_artist_language_accent_mapping` method completely as according to your suggestions.

Thanks.

Ashu463 committed:
fix: added test cases for newly created method in domain obj

Ashu463 committed:
fix: added test cases for line 402 and 429

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+                    % nested_dict
+                )
+            for language_code, (artist_id, voiceover_dict) in (
+            nested_dict.items()):
```
Fix indent.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+                        'Artist ID should not be empty.')
+                if not isinstance(voiceover_dict, dict):
+                    raise utils.ValidationError(
+                        'Expected voiceover_dict to be dict '
```
```suggestion
                        'Expected voiceover_dict to be a dict, '
```

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+                if not (
+                    utils.is_supported_audio_language_code(language_code)):
+                    raise utils.ValidationError(
+                        'Expected language code to be one from '
```
```suggestion
                        'Expected language code to be in '
```

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+                    utils.is_supported_audio_language_code(language_code)):
+                    raise utils.ValidationError(
+                        'Expected language code to be one from '
+                        'supported audio languages recieved: %s'
```
```suggestion
                        'supported audio languages, recieved: %s'
```

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+                    )
+                if not isinstance(artist_id, str):
+                    raise utils.ValidationError(
+                        'Expected artist id to be string recieved: %s'
```
```suggestion
                        'Expected artist id to be string, recieved: %s'
```

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+            nested_dict.items()):
+                if not isinstance(language_code, str):
+                    raise utils.ValidationError(
+                        'Expected language to be string recieved: %s'
```
```suggestion
                        'Expected language to be string, recieved: %s'
```

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+                raise utils.ValidationError('Content ID should not be empty.')
+            if not isinstance(nested_dict, dict):
+                raise utils.ValidationError(
+                    'Expected nested dict to be Dictionary, recieved: %s.'
```
```suggestion
                    'Expected nested dict to be a dict, recieved: %s.'
```

ankita240796 commented on core/domain/voiceover_services.py:
```diff
@@ -506,6 +506,58 @@ def get_autogeneratable_language_accent_list() -> Dict[str, Dict[str, str]]:
+
+
+def set_exploration_voice_artists_link(
+    exploration_voice_artists_link: voiceover_domain.ExplorationVoiceArtistsLink
```
If we pass in `exp_id` as a param here, we don't need `exp_id` in the domain object, right?

Ashu463 replied:
Yep you are right, we only need exp_id for set method.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+        """
+        result: Dict[str, Dict[str, str]] = {}
+        for lang_voiceover_mapping_tuple in (
+        self.content_id_to_voiceovers_mapping.values()):
```
```suggestion
            self.content_id_to_voiceovers_mapping.values()):
```

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+        self.content_id_to_voiceovers_mapping.values()):
+
+            for lang_code, voiceover_tuple in (
+            lang_voiceover_mapping_tuple.items()):
```
Ditto.
```suggestion
                lang_voiceover_mapping_tuple.items()):
```

Ashu463 replied:
I got confused in previous iteration abt this indent fixing, thanks for clearing it.

ankita240796 commented on core/domain/voiceover_services.py:
```diff
@@ -519,65 +571,40 @@ def get_autogeneratable_language_accent_codes() -> List[str]:
     return language_accent_codes
 
 
-def get_all_voice_artist_language_accent_mapping() -> Dict[str, Dict[str, str]]:
-    """The method returns a dict with all voice artist IDs as keys and nested
-    dicts as values. Each nested dict contains language codes as keys and
-    language accent codes as values.
+def get_voice_artist_metadata_mapping() -> Dict[str, Dict[str, str]]:
+    """Gets a mapping of voice artist IDs to language-accent mappings.
```
These changes are hard to read. Here is what I would recommend:

- Have a function for obtaining all domain objects i.e. fetch all models and then generate and return all domain objects.
- Have a method on domain object to extract `voice_artist_language_accent_mapping`
- In `get_all_voice_artist_language_accent_mapping`:
    - Obtain all domain objects
    - Get voice artist language accent mapping for each
    - Merge these mappings and return that
    - Do the same for `get_voice_artist_ids_to_voice_artist_names`

Does that make sense? If not, let's discuss before making any code changes and I can explain further.

Ashu463 replied:
Roger that ankita, here is the updated `get_all_voice_artist_language_accent_mapping()` method as per your suggestion. I'm waiting for its review if approved, then made similar changes with `get_voice_artist_ids_to_voice_artist_names`.

```python
def get_all_voice_artist_language_accent_mapping() -> Dict[str, Dict[str, str]]:
    """The method returns a dict with all voice artist IDs as keys and nested
    dicts as values. Each nested dict contains language codes as keys and
    language accent codes as values.
    Returns:
        dict(str, dict(str, str)). A dict representing voice artist IDs to
        language mappings.
    """
    voice_artist_id_to_language_mapping: Dict[str, Dict[str, str]] = (
        get_voice_artist_metadata_mapping())
    exp_voice_artist_link: List[voiceover_domain.ExplorationVoiceArtistsLink] = (
        get_exploration_voice_artists_links()
    )

    all_voice_artist_to_language_mapping :Dict[str, Dict[str, str]] = {}
    for link in exp_voice_artist_link:
        link_mappings = link.get_voice_artist_language_mappings(
            voice_artist_id_to_language_mapping)
        
        for voice_artist_id, language_mapping in link_mappings.items():
            if voice_artist_id not in all_voice_artist_to_language_mapping:
                all_voice_artist_to_language_mapping[voice_artist_id] = {}
            all_voice_artist_to_language_mapping[voice_artist_id].update(language_mapping)
    
    return all_voice_artist_to_language_mapping

```

ankita240796 replied:
Thanks @Ashu463, can you elaborate what does `get_voice_artist_metadata_mapping` do and why is it required?

Ashu463 replied:
Yeah sure, get_voice_artist_metadata_mapping returns a reordered dict by fetching data from `VoiceArtistMetadataModel` and format of this reorder is dict of ```model.id: model.language_code_to_accent``` .
It is required to get all langauge_codes along with their accents associated with a voiceover artist.

ankita240796 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+
+    def get_voice_artist_language_mappings(
+        self,
+        voice_artist_id_to_language_mapping: Dict[str, Dict[str, str]]
```
Unclear on what is the usecase of this argument. Can you elaborate?

Ashu463 replied:
It is used to check whether the voice_artist_id and lang_code from ExplorationVoiceArtistsLink model present in VoiceArtistMetadataModel or not and hence to get accent of languages codes for mapping.

ankita240796 replied:
Hmm, so the comment I made on merging the mappings in voiceover services removes the need for this, right? Basically let each domain object only return it's own mapping and then handle the merging in services. Does that make sense?

Ashu463 replied:
Hey you said to make a new method in domain object to fetch `voice_artist_language_accent_mapping` at second point, and it can't be removed either it will present of domain object side or service side. 
This method returns a combination VoiceArtistMetadataModel and ExplorationVoiceArtistsLinkmodel, and merging is done on the service file, I'm lil bit confused with your cmnt. It's a good practice to return domain object's own mapping but that couldn't be possible in this case, else I need to move this logic to service side.

ankita240796 replied:
Hmm okay, there is some confusion here. Let me list down what's happening:

We are generating a mapping of `<artist_id: <language_code, <accent codes>>>`. To generate this, we need two models `VoiceArtistMetadataModel` and `ExplorationVoiceArtistsLinkmodel`. My understanding was that merging could entirely be done in service file with the `ExplorationVoiceArtistsLink` domain object just returning a `<lang code, voice artist IDs>` mapping.

So, in domain object:

```
def get_lang_code_to_artist_id_mapping(self):
    content_id_to_voiceovers_mapping = (
        exp_voice_artist_model.content_id_to_voiceovers_mapping)

    lang_code_to_artist_id = {} # Use appropriate type

    for lang_voiceover_mapping_tuple in (
        content_id_to_voiceovers_mapping.values()):

        for lang_code, voiceover_tuple in (
            lang_voiceover_mapping_tuple.items()):

             voice_artist_id = voiceover_tuple[0]
             lang_code_to_artist_id[lang_code] = voice_artist_id

return lang_code_to_artist_ids
```

Can't this mapping be then used in the services to obtain the final accent code based mapping?

Ashu463 replied:
Cool, everything is cleared now. I updated voiceover_domain as per your suggestion, merging will now done in voiceover_services and specifically this method would return Dict of lang_code and voice_artist_id.

ankita240796 commented on core/domain/voiceover_services.py:
```diff
@@ -894,17 +942,15 @@ def update_exploration_voice_artist_link_model(
                 content_id_to_voiceovers_mapping[content_id][lang_code] = (
                     user_id, voiceover_dict.to_dict())
 
-    if exp_voice_artist_link_model is None:
-        exp_voice_artist_link_model = (
-            create_exploration_voice_artists_link_model_instance(
-                exploration_id=updated_exploration.id,
+    if exploration_voice_artists_link is None:
+        exploration_voice_artists_link = (
+            voiceover_domain.ExplorationVoiceArtistsLink(
+                exp_id=updated_exploration.id,
```
Revert changes in this method as it to update the model and you do not need to do the extra conversion from model to domain object and back to the model for that. Domain object should be provided by voiceover_services to anywhere else in the codebase instead of providing the model directly but internally voiceover_services can access and update the model.

Ashu463 replied:
Totally agree with your points, but tell me one thing do we need to validate data before sending it to model ? If not then that will end up whole use case of `set_exploration_voice_artists_link` method and we can remove `exp_id` from domain obj even.

ankita240796 replied:
> Totally agree with your points, but tell me one thing do we need to validate data before sending it to model ?

Can you explain what you mean by this? Do you mean should we validate data before saving it to datastore - the answer is yes, we should.

However, even if we do that, why is `exp_id` required then?

Ashu463 replied:
Hey I agreed at your **do not need to do extra conversions** point, and then asked that question of validating data before sending it storage layer, which you answered yes, now that's the point I implemented. 
Now for exp_id: there may be the cases when someone updates the exploration_id, and before updating it to storage layer we first need to validate it, so yeah there is necessity of keeping this.

ankita240796 commented on core/domain/voiceover_services.py:
```diff
@@ -815,12 +858,18 @@ def create_exploration_voice_artists_link_model_instance(
             content_id_to_voiceovers_mapping=content_id_to_voiceovers_mapping
         )
     )
+    exploration_voice_artists_link.validate()
+    exploration_voice_artists_link_model = (
+        voiceover_models.ExplorationVoiceArtistsLinkModel(
+        id=exploration_id,
+        content_id_to_voiceovers_mapping=content_id_to_voiceovers_mapping
+    ))
```
Ditto - same for this method.

ankita240796 reviewed (changes requested):
Thanks @Ashu463, PTAL.

Ashu463 commented:
@ankita240796 left comments on your suggestion need some discussion there, so it is just a ping not asking reivew now.

Thanks.

Ashu463 committed:
fix: iteration#3

ankita240796 reviewed (commented):
Thanks @Ashu463, haven't taken a full pass but responded to your queries as asked.

ankita240796 commented:
Hi @Ashu463, just a note: please assign the PR to me if you have any questions for me? That will make sure I do not end up missing it. Thanks!

Nik-09 commented on core/domain/voiceover_domain.py:
```diff
@@ -352,3 +352,144 @@ def create_empty(
+                'Expected content_id_to_voiceovers_mapping to be Dict'
+                'But recieved: %s' % self.content_id_to_voiceovers_mapping
+            )
+        for content_id, nested_dict in (
```
Can we use a more meaningful name instead of nested dict?

Ashu463 replied:
Yeah sure, is `langauge_to_voiceover_dict` is fine ?

ankita240796 replied:
Yes, sounds good, let's use that

Ashu463 replied:
Done.

Ashu463 commented:
@ankita240796 friendly ping for this.

ankita240796 commented:
@Ashu463 please assign the reviewer to the PR when you want a re-review as per the last point under. https://github.com/oppia/oppia/wiki/Rules-for-making-PRs#step-5-address-review-comments-until-all-reviewers-approve. Thanks!

ankita240796 commented on core/controllers/voiceover.py:
```diff
@@ -168,7 +168,6 @@ def get(self) -> None:
             voiceover_services.get_all_voice_artist_language_accent_mapping())
         voice_artist_id_to_voice_artist_name = (
             voiceover_services.get_voice_artist_ids_to_voice_artist_names())
-
```
Remove - unintended change.

Ashu463 committed:
Merge branch 'develop' into issue#21970-ExplorationVoiceArtistsLink

Ashu463 committed:
Merge branch 'develop' into issue#21970-ExplorationVoiceArtistsLink

Ashu463 committed:
fix: updated voiceover domain method

Ashu463 committed:
Merge branch 'develop' into issue#21970-ExplorationVoiceArtistsLink

Ashu463 committed:
fix: unintended changes on voiceovers

Ashu463 committed:
fix: updated test cases of voiceover domain

Ashu463 commented:
@ankita240796 I made the changes as per your suggestion. It's been a long time that I looked on this PR, so will explain what changes I done and why for your recall. 
- Created domain object and wrote validation for content_id_to_voiceovers_mapping. Voiceover dict within the tuple is validated by state_domain.voiceover's validate method, which is enclosed in try-except block to shoot any unexpected error.
- Created test file for the domain object which is reviewed till now.
- Changes in voiceover_service file: 
  - Created get and save methods to get and save this model. And incoming data is validated before saving it storage models.
  - Refactored `get_all_voice_artist_language_accent_mapping` as per your discussion. 
  
So now, this PR is ready for complete review, and hoping it will be ready to merge after this iteration. 
Thanks.

lkbhitesh07 commented on core/domain/voiceover_domain.py:
```diff
@@ -316,3 +316,122 @@ def create_empty(
+    """Domain object for linking Exploration with voice artists."""
+
+    def __init__(
+            self,
```
Indentation missing here.

Ashu463 replied:
Hmm, done.

lkbhitesh07 reviewed (changes requested):
Some nits, Also, I am wondering what will happen to the existing data once we try to invoke the domain validate method? Also, before saving the data into the datastore model I think we should call the validate method.
Thanks

Ashu463 committed:
Merge branch 'develop' into issue#21970-ExplorationVoiceArtistsLink

Ashu463 committed:
fix: removed expected and missing validation keys
